src/body-data-wrapper.js

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+/* eslint-disable no-param-reassign */
+const { Request } = require('@adobe/helix-fetch');
+
+/**
+ * Extracts the _data_ from the given request. The data can be provided either as request
+ * parameters, url-encoded form data body, or a json body.
+ *
+ * Note that for post body requests, the body is consumed from the request and is no longer
+ * available.
+ *
+ * @param {Request} request The universal request
+ * @param {BodyDataOptions} [opts] Options
+ * @returns {Promise<object>} the parsed data object.
+ */
+async function getData(request, opts) {
+  const { coerceNumber, coerceInt, coerceBoolean } = opts;
+  if (/json/.test(request.headers.get('content-type'))) {
+    // eslint-disable-next-line no-return-await
+    return await request.json();
+  }
+  let data = new URL(request.url).searchParams;
+  if (/^application\/x-www-form-urlencoded/.test(request.headers.get('content-type'))) {
+    // todo: combine urlencoded and request params
+    data = new URLSearchParams(await request.text());
+  }
+  return Array.from(data.entries()).reduce((alldata, [key, value]) => {
+    const bracketpattern = /\[([0-9]*)\]$/;
+    // check for key names like [1] or [0]
+    const isArray = bracketpattern.test(key);
+    const arrayIndex = Number.parseInt(isArray && key.match(bracketpattern)[1], 10);
+    const cleanKey = key.replace(bracketpattern, '');
+
+    const values = (isArray ? [value] : data.getAll(key))
+      .map((v) => (coerceBoolean && (v === 'true' || v === 'false') ? v === 'true' : v))
+      .map((v) => (coerceInt && /^-?[0-9]+$/.test(v) ? Number.parseInt(v, 10) : v))
```

Comment:
there is no difference between integer and number... why would it matter?

Reply:
In JS, there isn't a difference, but in practice there is. If you want to get a Fastly Service ID version number, you know that it is an integer, an `version = 11.00000000000001` could be created by your client-side code, but should not be accepted if you really want integers.

Reply:
but then, the flags are too broad. what if you have numbers and _integers_ in your post data....
I would remove all coercing and let the function decide....

Reply:
If you have numbers and integers, use both flags. As integers are a subset of numbers, they run first.

Reply:
but your example with `11.000001` wouldn't work with coerce integers. as it fails the test.....
I would just remove all number magic here....